sdk/keyvault/keyvault-admin/src/accessControlClient.ts

```diff
@@ -0,0 +1,442 @@
+   *
+   * Example usage:
+   * ```ts
+   * import { AccessControlClient } from "@azure/keyvault-admin";
+   * import { DefaultAzureCredential } from "@azure/identity";
+   *
+   * let vaultUrl = `https://<MY KEYVAULT HERE>.vault.azure.net`;
+   * let credentials = new DefaultAzureCredential();
+   *
+   * let client = new AccessControlClient(vaultUrl, credentials);
+   * ```
+   * @param {string} vaultUrl the URL of the Key Vault. It should have this shape: https://${your-key-vault-name}.vault.azure.net
+   * @param {TokenCredential} credential An object that implements the `TokenCredential` interface used to authenticate requests to the service. Use the @azure/identity package to create a credential that suits your needs.
+   * @param {AccessControlClientOptions} [pipelineOptions] Pipeline options used to configure Key Vault API requests. Omit this parameter to use the default pipeline configuration.
+   * @memberof AccessControlClient
+   */
+  constructor(
+    vaultUrl: string,
+    credential: TokenCredential,
+    pipelineOptions: AccessControlClientOptions = {}
+  ) {
+    this.vaultUrl = vaultUrl;
+
+    const libInfo = `azsdk-js-keyvault-admin/${SDK_VERSION}`;
+
+    const userAgentOptions = pipelineOptions.userAgentOptions;
+
+    pipelineOptions.userAgentOptions = {
+      ...pipelineOptions.userAgentOptions,
+      userAgentPrefix:
+        userAgentOptions && userAgentOptions.userAgentPrefix
+          ? `${userAgentOptions.userAgentPrefix} ${libInfo}`
+          : libInfo
+    };
+
+    const authPolicy = isTokenCredential(credential)
+      ? challengeBasedAuthenticationPolicy(credential)
+      : signingPolicy(credential);
+
+    const internalPipelineOptions = {
+      ...pipelineOptions,
+      ...{
+        loggingOptions: {
+          logger: logger.info,
+          logPolicyOptions: {
+            allowedHeaderNames: [
+              "x-ms-keyvault-region",
+              "x-ms-keyvault-network-info",
+              "x-ms-keyvault-service-version"
+            ]
+          }
+        }
+      }
+    };
+
+    const pipeline = createPipelineFromOptions(internalPipelineOptions, authPolicy);
+    this.client = new KeyVaultClient({
+      apiVersion: pipelineOptions.serviceVersion || LATEST_API_VERSION,
+      ...pipeline
+    });
+  }
+
+  /**
+   * Creates a role assignment in an Azure Key Vault.
+   *
+   * Example usage:
+   * ```ts
+   * const client = new AccessControlClient(url, credentials);
+   * const result = await client.createRoleAssignment("/", "295c179b-9ad3-4117-99cd-b1aa66cf4517");
+   * ```
+   * @summary Creates a new role assignment.
+   * @param {RoleAssignmentScope} scope The scope of the role assignment.
+   * @param {string} name The name of the role assignment. Must be a UUID.
+   * @param {CreateRoleAssignmentOptions} [options] The optional parameters.
+   */
+  public async createRoleAssignment(
+    scope: RoleAssignmentScope,
+    name: string,
+    options?: CreateRoleAssignmentOptions
```

Comment:
The `RoleAssignmentProperties` are required

Reply:
Ohh! thank you. I'll update this.

Reply:
Are you sure about this? The generated code shows that in the RoleAssignmentProperties (even the one with the scope) all of the properties are optional. Doesn't that make the properties parameter optional? At least on the convenience layer.

Reply:
The `RoleAssignmentProperties` is where the `principalId` and `roleDefinitionId` are specified. Is it possible to create a role assignment without those?

Reply:
I have no idea yet, but the generated code sure doesn't enforce this: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-admin/src/generated/models/index.ts#L133-L146

Reply:
That is to say, if these are required: It would make sense for me to make an interface with principalId and roleDefinitionId as required properties, and not optional properties.

Reply:
Yes, I believe so since the request will fail without them.

Reply:
Gotcha! Thank you. I'll update this.

Reply:
I wonder if we should separate these parameters from the options bag into parameters of this specific method. What do you think, @heaths ?

Reply:
Generators may have bugs or haven't implemented features yet. E.g. .NET's just implemented x-nullable recently. Swagger should be your authoritative source:

1. RoleAssignmentProperties: https://github.com/Azure/azure-rest-api-specs/blob/6703e4c8b4526b9cb8e563afa156756ca89e723d/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.2-preview/rbac.json#L351 (both properties required)
2. RoleAssignmentCreateParameters: https://github.com/Azure/azure-rest-api-specs/blob/6703e4c8b4526b9cb8e563afa156756ca89e723d/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.2-preview/rbac.json#L368 (requires the aforementioned properties)

So both are required and should be method parameters.

Reply:
Ok so I got from Heath that: These two parameters should be required by the method, as individual parameters.

Reply:
Addressed here: https://github.com/Azure/azure-sdk-for-js/pull/10815/commits/b7d3bf58cb6442a2e87b7ec2c780ab6181844474

Please review again!